clippy_lints/src/len_zero.rs

```diff
@@ -202,39 +202,111 @@ fn check_trait_items(cx: &LateContext<'_>, visited_trait: &Item<'_>, trait_items
     }
 }
 
-fn check_impl_items(cx: &LateContext<'_>, item: &Item<'_>, impl_items: &[ImplItemRef<'_>]) {
-    fn is_named_self(cx: &LateContext<'_>, item: &ImplItemRef<'_>, name: &str) -> bool {
-        item.ident.name.as_str() == name
-            && if let AssocItemKind::Fn { has_self } = item.kind {
-                has_self && cx.tcx.fn_sig(item.id.def_id).inputs().skip_binder().len() == 1
-            } else {
-                false
-            }
-    }
+fn check_impl_items(cx: &LateContext<'_>, item: &Item<'_>) {
+    let len = Symbol::intern("len");
+    let is_empty = Symbol::intern("is_empty");
+    let (len, is_empty) = cx
```

Comment:
Can we only start this check if we find a `len()` method on the AST?

It's typical for lints to have performant quick-discard checks first, since they get run on many many nodes.

I would structure this as a `check_impl_item`, where we immediately bail for non-`len` impl items, and if we find a `len` then we perform the `inherent_impls` search, which can be more expensive.

Reply:
What is the performance of the `inherent_impls` query, given that it should be cached at this point. After that it's a loop over an array with two hash table lookups each iteration. This is done once per type type.

Compared with a symbol comparison per `ImplItem`, I don't see it being much slower, if it even is. Really it would come down to the number of `ImplItem`s compared to the number of types declared.

Reply:
Yeah that's fair. I think the important thing is that this introduces a loop, whereas the loop over "every impl item" already happens in the lint engine. The general rule of thumb to follow is to avoid new loops without some fast-discard check.

I suspect the perf difference will not be huge (but it adds up). I do think this will make the code cleaner as well though.

Reply:
The loop would be zero or one iterations in the normal case. It's over each non-trait impl block for the type, not each item. Going the other route will also be looking at all `ImplItem`s (traits and inherent).

Reply:
Correct, what I'm trying to say is that "looking at all ImplItems" is an operation that is already happening due to how the linting architecture works -- every lint has every method called for every AST node (with some optimizations), even if those methods are empty.

In general the table lookups often have noticeable perf impact, even if they're hashtables, due to cache issues.

Reply:
It might be a little clearer, just by virtue of separating the search for the `len` function, from the search for the `is_empty` function. This approach does mean multiple `len` functions on the same type won't trigger the lint multiple times (yes, I know how ridiculous of an edge case this is).

Reply:
That is a good point, there's a good chance of a cache miss.

Reply:
Done.